transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDatagramChannel.java

```diff
@@ -567,4 +528,142 @@ void epollInReady() {
+            InetSocketAddress local = localAddress();
+            if (received == 1) {
+                // Single packet fast-path
+                DatagramPacket packet = packets[0].newDatagramPacket(byteBuf, local);
+                allocHandle.lastBytesRead(datagramSize);
+                allocHandle.incMessagesRead(1);
+                pipeline().fireChannelRead(packet);
+                byteBuf = null;
+                return true;
+            }
+
+            // Its important that we process all received data out of the NativeDatagramPacketArray
+            // before we call fireChannelRead(...). This is because the user may call flush()
+            // in a channelRead(...) method and so may re-use the NativeDatagramPacketArray again.
+            bufferPackets = RecyclableArrayList.newInstance();
+            for (int i = 0; i < received; i++) {
+                DatagramPacket packet = packets[i].newDatagramPacket(byteBuf.readRetainedSlice(datagramSize), local);
+                bufferPackets.add(packet);
+            }
+
+            allocHandle.lastBytesRead(datagramSize * received);
+            allocHandle.incMessagesRead(received);
+
+            for (int i = 0; i < received; i++) {
+                pipeline().fireChannelRead(bufferPackets.set(i, Unpooled.EMPTY_BUFFER));
+            }
+            bufferPackets.recycle();
+            bufferPackets = null;
+            return true;
+        } finally {
+            if (byteBuf != null) {
+                byteBuf.release();
+            }
+            if (bufferPackets != null) {
+                for (int i = 0; i < bufferPackets.size(); i++) {
+                    ReferenceCountUtil.release(bufferPackets.get(i));
+                }
+                bufferPackets.recycle();
+            }
+        }
+    }
+
+    private boolean read(EpollRecvByteAllocatorHandle allocHandle, ByteBuf byteBuf) throws IOException {
+        try {
+            allocHandle.attemptedBytesRead(byteBuf.writableBytes());
+
+            final DatagramSocketAddress remoteAddress;
+            if (byteBuf.hasMemoryAddress()) {
+                // has a memory address so use optimized call
+                remoteAddress = socket.recvFromAddress(
+                        byteBuf.memoryAddress(), byteBuf.writerIndex(), byteBuf.capacity());
+            } else {
+                ByteBuffer nioData = byteBuf.internalNioBuffer(
+                        byteBuf.writerIndex(), byteBuf.writableBytes());
+                remoteAddress = socket.recvFrom(nioData, nioData.position(), nioData.limit());
+            }
+
+            if (remoteAddress == null) {
+                allocHandle.lastBytesRead(-1);
+                return false;
+            }
+            InetSocketAddress localAddress = remoteAddress.localAddress();
+            if (localAddress == null) {
+                localAddress = localAddress();
+            }
+            allocHandle.lastBytesRead(remoteAddress.receivedAmount());
```

Comment:
Technically we should also pass the same value as was passed to `attemptedBytesRead` here in the case that scattering is enabled... since it could also take this path if the current recv buffer size only fits one packet. And if we then report that not all of the buffer was used, maybe it will never grow larger again in adaptive case (though I didn't look in detail at the adaptive algorithm). Or could be circumvented by setting the min size of `AdaptiveRecvByteBufAllocator` to at least 2x the max datagram size (then would never take this path).

Reply:
fixed